File: library/src/main/java/com/github/sundeepk/compactcalendarview/CompactCalendarView.java
package com.github.sundeepk.compactcalendarview;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.v4.view.GestureDetectorCompat;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.VelocityTracker;
import android.view.View;
import android.widget.OverScroller;

import com.github.sundeepk.compactcalendarview.domain.Event;

import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

public class CompactCalendarView extends View {

    public static final int FILL_LARGE_INDICATOR = 1;
    public static final int NO_FILL_LARGE_INDICATOR = 2;
    public static final int SMALL_INDICATOR = 3;

    private final AnimationHandler animationHandler;
    private CompactCalendarController compactCalendarController;
    private GestureDetectorCompat gestureDetector;
    private boolean horizontalScrollEnabled = true;

    public interface CompactCalendarViewListener {
        public void onDayClick(Date dateClicked);
        public void onMonthScroll(Date firstDayOfNewMonth);
    }

    public interface CompactCalendarAnimationListener {
        public void onOpened();
        public void onClosed();
    }

    private final GestureDetector.SimpleOnGestureListener gestureListener = new GestureDetector.SimpleOnGestureListener() {
        @Override
        public void onLongPress(MotionEvent e) {
        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            compactCalendarController.onSingleTapUp(e);
            invalidate();
            return super.onSingleTapUp(e);
        }

        @Override
        public boolean onDown(MotionEvent e) {
            return true;
        }

        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            return true;
        }

        @Override
        public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
            if(horizontalScrollEnabled) {
                if (Math.abs(distanceX) > 0) {
                    getParent().requestDisallowInterceptTouchEvent(true);

                    compactCalendarController.onScroll(e1, e2, distanceX, distanceY);
                    invalidate();
                    return true;
                }
            }

            return false;
        }
    };

    public CompactCalendarView(Context context) {
        this(context, null);
    }

    public CompactCalendarView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public CompactCalendarView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        compactCalendarController = new CompactCalendarController(new Paint(), new OverScroller(getContext()),
                new Rect(), attrs, getContext(),  Color.argb(255, 233, 84, 81),
                Color.argb(255, 64, 64, 64), Color.argb(255, 219, 219, 219), VelocityTracker.obtain(),
                Color.argb(255, 100, 68, 65), new EventsContainer(Calendar.getInstance()),
                Locale.getDefault(), TimeZone.getDefault());
        gestureDetector = new GestureDetectorCompat(getContext(), gestureListener);
        animationHandler = new AnimationHandler(compactCalendarController, this);
    }

    public void setAnimationListener(CompactCalendarAnimationListener compactCalendarAnimationListener){
        animationHandler.setCompactCalendarAnimationListener(compactCalendarAnimationListener);
    }

    /*
    Use a custom locale for compact calendar and reinitialise the view.
     */
    public void setLocale(TimeZone timeZone, Locale locale){
        compactCalendarController.setLocale(timeZone, locale);
        invalidate();
    }

    /*
    Compact calendar will use the locale to determine the abbreviation to use as the day column names.
    The default is to use the default locale and to abbreviate the day names to one character.
    Setting this to true will displace the short weekday string provided by java.
     */
    public void setUseThreeLetterAbbreviation(boolean useThreeLetterAbbreviation){
        compactCalendarController.setUseWeekDayAbbreviation(useThreeLetterAbbreviation);
        invalidate();
    }

    public void setCalendarBackgroundColor(final int calenderBackgroundColor) {
        compactCalendarController.setCalenderBackgroundColor(calenderBackgroundColor);
        invalidate();
    }

    /*
    Sets the name for each day of the week. No attempt is made to adjust width or text size based on the length of each day name.
    Works best with 3-4 characters for each day.
     */
    public void setDayColumnNames(String[] dayColumnNames){
        compactCalendarController.setDayColumnNames(dayColumnNames);
    }

    public void setFirstDayOfWeek(int dayOfWeek){
        compactCalendarController.setFirstDayOfWeek(dayOfWeek);
        invalidate();
    }

    public void setCurrentSelectedDayBackgroundColor(int currentSelectedDayBackgroundColor) {
        compactCalendarController.setCurrentSelectedDayBackgroundColor(currentSelectedDayBackgroundColor);
        invalidate();
    }

    public void setCurrentDayBackgroundColor(int currentDayBackgroundColor) {
        compactCalendarController.setCurrentDayBackgroundColor(currentDayBackgroundColor);
        invalidate();
    }

    public int getHeightPerDay(){
        return compactCalendarController.getHeightPerDay();
    }

    public void setListener(CompactCalendarViewListener listener){
        compactCalendarController.setListener(listener);
    }

    public Date getFirstDayOfCurrentMonth(){
        return compactCalendarController.getFirstDayOfCurrentMonth();
    }

    public void shouldDrawIndicatorsBelowSelectedDays(boolean shouldDrawIndicatorsBelowSelectedDays){
        compactCalendarController.shouldDrawIndicatorsBelowSelectedDays(shouldDrawIndicatorsBelowSelectedDays);
    }

    public void setCurrentDate(Date dateTimeMonth){
        compactCalendarController.setCurrentDate(dateTimeMonth);
        invalidate();
    }

    public int getWeekNumberForCurrentMonth(){
        return compactCalendarController.getWeekNumberForCurrentMonth();
    }

    public void setShouldDrawDaysHeader(boolean shouldDrawDaysHeader){
        compactCalendarController.setShouldDrawDaysHeader(shouldDrawDaysHeader);
    }

    public void setCurrentSelectedDayTextColor(int currentSelectedDayTextColor) {
        compactCalendarController.setCurrentSelectedDayTextColor(currentSelectedDayTextColor);
    }

    public void setCurrentDayTextColor(int currentDayTextColor) {
        compactCalendarController.setCurrentDayTextColor(currentDayTextColor);
    }

    /**
     * see {@link #addEvent(Event, boolean)} when adding single events
     * or {@link #addEvents(java.util.List)}  when adding multiple events
     * @param event
     */
    @Deprecated
    public void addEvent(Event event){
        addEvent(event, false);
    }

    /**
     *  Adds an event to be drawn as an indicator in the calendar.
     *  If adding multiple events see {@link #addEvents(List)}} method.
     * @param event to be added to the calendar
     * @param shouldInvalidate true if the view should invalidate
     */
    public void addEvent(Event event, boolean shouldInvalidate){
        compactCalendarController.addEvent(event);
        if(shouldInvalidate){
            invalidate();
        }
    }

    /**
     * Adds multiple events to the calendar and invalidates the view once all events are added.
     */
    public void addEvents(List<Event> events){
        compactCalendarController.addEvents(events);
        invalidate();
    }

    /**
     * Fetches the events for the date passed in
     * @param date
     * @return
     */
    public List<Event> getEvents(Date date){
        return compactCalendarController.getCalendarEventsFor(date.getTime());
    }

    /**
     * Fetches the events for the epochMillis passed in
     * @param epochMillis
     * @return
     */
    public List<Event> getEvents(long epochMillis){
        return compactCalendarController.getCalendarEventsFor(epochMillis);
    }

    /**
     * Fetches the events for the month of the epochMillis passed in and returns a sorted list of events
     * @param epochMillis
     * @return
     */
    public List<Event> getEventsForMonth(long epochMillis){
        return compactCalendarController.getCalendarEventsForMonth(epochMillis);
    }

    /**
     * Fetches the events for the month of the date passed in and returns a sorted list of events
     * @param date
     * @return
     */
    public List<Event> getEventsForMonth(Date date){
        return compactCalendarController.getCalendarEventsForMonth(date.getTime());
    }

    /**
     * Remove the event associated with the Date passed in
     * @param date
     */
    public void removeEvents(Date date){
        compactCalendarController.removeEventsFor(date.getTime());
    }

    public void removeEvents(long epochMillis){
        compactCalendarController.removeEventsFor(epochMillis);
    }

    /**
     * see {@link #removeEvent(Event, boolean)} when removing single events
     * or {@link #removeEvents(java.util.List)} (java.util.List)}  when removing multiple events
     * @param event
     */
    @Deprecated
    public void removeEvent(Event event){
        removeEvent(event, false);
    }

    /**
     * Removes an event from the calendar.
     * If removing multiple events see {@link #removeEvents(List)}
     *
     * @param event event to remove from the calendar
     * @param shouldInvalidate true if the view should invalidate
     */
    public void removeEvent(Event event, boolean shouldInvalidate){
        compactCalendarController.removeEvent(event);
        if(shouldInvalidate){
            invalidate();
        }
    }

    /**
     * Removes multiple events from the calendar and invalidates the view once all events are added.
     */
    public void removeEvents(List<Event> events){
        compactCalendarController.removeEvents(events);
        invalidate();
    }

    /**
     * Clears all Events from the calendar.
     */
    public void removeAllEvents() {
        compactCalendarController.removeAllEvents();
        invalidate();
    }

    public void shouldSelectFirstDayOfMonthOnScroll(boolean shouldSelectFirstDayOfMonthOnScroll){
        compactCalendarController.setShouldSelectFirstDayOfMonthOnScroll(shouldSelectFirstDayOfMonthOnScroll);
    }

    public void setCurrentSelectedDayIndicatorStyle(final int currentSelectedDayIndicatorStyle){
        compactCalendarController.setCurrentSelectedDayIndicatorStyle(currentSelectedDayIndicatorStyle);
        invalidate();
    }

    public void setCurrentDayIndicatorStyle(final int currentDayIndicatorStyle){
        compactCalendarController.setCurrentDayIndicatorStyle(currentDayIndicatorStyle);
        invalidate();
    }

    public void setEventIndicatorStyle(final int eventIndicatorStyle){
        compactCalendarController.setEventIndicatorStyle(eventIndicatorStyle);
        invalidate();
    }

    private void checkTargetHeight() {
        if (compactCalendarController.getTargetHeight() <= 0) {
            throw new IllegalStateException("Target height must be set in xml properties in order to expand/collapse CompactCalendar.");
        }
    }

    public void displayOtherMonthDays(boolean displayOtherMonthDays){
        compactCalendarController.setDisplayOtherMonthDays(displayOtherMonthDays);
        invalidate();
    }

    public void setTargetHeight(int targetHeight){
        compactCalendarController.setTargetHeight(targetHeight);
        checkTargetHeight();
    }

    public void showCalendar(){
        checkTargetHeight();
        animationHandler.openCalendar();
    }

    public void hideCalendar(){
        checkTargetHeight();
        animationHandler.closeCalendar();
    }

    public void showCalendarWithAnimation(){
        checkTargetHeight();
        animationHandler.openCalendarWithAnimation();
    }

    public void hideCalendarWithAnimation(){
        checkTargetHeight();
        animationHandler.closeCalendarWithAnimation();
    }

    public void showNextMonth(){
        compactCalendarController.showNextMonth();
        invalidate();
    }

    public void showPreviousMonth(){
        compactCalendarController.showPreviousMonth();
        invalidate();
    }

    public boolean isAnimating(){
        return animationHandler.isAnimating();
    }

    @Override
    protected void onMeasure(int parentWidth, int parentHeight) {
        super.onMeasure(parentWidth, parentHeight);
        int width = MeasureSpec.getSize(parentWidth);
        int height = MeasureSpec.getSize(parentHeight);
        if(width > 0 && height > 0) {
            compactCalendarController.onMeasure(width, height, getPaddingRight(), getPaddingLeft());
        }
        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        compactCalendarController.onDraw(canvas);
    }

    @Override
    public void computeScroll() {
        super.computeScroll();
        if(compactCalendarController.computeScroll()){
            invalidate();
        }
    }

    public void shouldScrollMonth(boolean enableHorizontalScroll){
        this.horizontalScrollEnabled = enableHorizontalScroll;
    }

    public boolean onTouchEvent(MotionEvent event) {
        if (horizontalScrollEnabled) {
            compactCalendarController.onTouch(event);
            invalidate();
        }

        // on touch action finished (CANCEL or UP), we re-allow the parent container to intercept touch events (scroll inside ViewPager + RecyclerView issue #82)
        if((event.getAction() == MotionEvent.ACTION_CANCEL || event.getAction() == MotionEvent.ACTION_UP) && horizontalScrollEnabled) {
            getParent().requestDisallowInterceptTouchEvent(false);
        }

        // always allow gestureDetector to detect onSingleTap and scroll events
        return gestureDetector.onTouchEvent(event);
    }

    @Override
    public boolean canScrollHorizontally(int direction) {
        if (this.getVisibility() == View.GONE) {
            return false;
        }
        // Prevents ViewPager from scrolling horizontally by announcing that (issue #82)
        return this.horizontalScrollEnabled;
    }

}

File: library/src/main/java/com/github/sundeepk/compactcalendarview/CompactCalendarView.kt
package com.github.sundeepk.compactcalendarview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.VelocityTracker
import android.view.View
import android.widget.OverScroller
import androidx.core.view.GestureDetectorCompat
import com.github.sundeepk.compactcalendarview.domain.Event
import java.util.*
import kotlin.math.abs

class CompactCalendarView @JvmOverloads constructor(context: Context?, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {
    private val animationHandler: AnimationHandler
    private val compactCalendarController: CompactCalendarController = CompactCalendarController(Paint(), OverScroller(getContext()),
            Rect(), attrs, getContext(), Color.argb(255, 233, 84, 81),
            Color.argb(255, 64, 64, 64), Color.argb(255, 219, 219, 219), VelocityTracker.obtain(),
            Color.argb(255, 100, 68, 65), EventsContainer(Calendar.getInstance()),
            Locale.getDefault(), TimeZone.getDefault())
    private val gestureDetector: GestureDetectorCompat
    private var horizontalScrollEnabled = true

    interface CompactCalendarViewListener {
        fun onDayClick(dateClicked: Date?)
        fun onMonthScroll(firstDayOfNewMonth: Date?)
    }

    interface CompactCalendarAnimationListener {
        fun onOpened()
        fun onClosed()
    }

    private val gestureListener: SimpleOnGestureListener = object : SimpleOnGestureListener() {
        override fun onLongPress(e: MotionEvent) {}
        override fun onSingleTapUp(e: MotionEvent): Boolean {
            compactCalendarController.onSingleTapUp(e)
            invalidate()
            return super.onSingleTapUp(e)
        }

        override fun onDown(e: MotionEvent): Boolean {
            return true
        }

        override fun onFling(e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
            return true
        }

        override fun onScroll(e1: MotionEvent, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
            if (horizontalScrollEnabled) {
                if (abs(distanceX) > 0) {
                    parent.requestDisallowInterceptTouchEvent(true)
                    compactCalendarController.onScroll(e1, e2, distanceX, distanceY)
                    invalidate()
                    return true
                }
            }
            return false
        }
    }

    fun setAnimationListener(compactCalendarAnimationListener: CompactCalendarAnimationListener?) {
        animationHandler.setCompactCalendarAnimationListener(compactCalendarAnimationListener)
    }

    /*
    Use a custom locale for compact calendar and reinitialise the view.
     */
    fun setLocale(timeZone: TimeZone?, locale: Locale?) {
        compactCalendarController.setLocale(timeZone, locale)
        invalidate()
    }

    /*
    Compact calendar will use the locale to determine the abbreviation to use as the day column names.
    The default is to use the default locale and to abbreviate the day names to one character.
    Setting this to true will displace the short weekday string provided by java.
     */
    fun setUseThreeLetterAbbreviation(useThreeLetterAbbreviation: Boolean) {
        compactCalendarController.setUseWeekDayAbbreviation(useThreeLetterAbbreviation)
        invalidate()
    }

    fun setCalendarBackgroundColor(calenderBackgroundColor: Int) {
        compactCalendarController.setCalenderBackgroundColor(calenderBackgroundColor)
        invalidate()
    }

    /*
    Sets the name for each day of the week. No attempt is made to adjust width or text size based on the length of each day name.
    Works best with 3-4 characters for each day.
     */
    fun setDayColumnNames(dayColumnNames: Array<String?>?) {
        compactCalendarController.setDayColumnNames(dayColumnNames)
    }

    fun setFirstDayOfWeek(dayOfWeek: Int) {
        compactCalendarController.setFirstDayOfWeek(dayOfWeek)
        invalidate()
    }

    fun setCurrentSelectedDayBackgroundColor(currentSelectedDayBackgroundColor: Int) {
        compactCalendarController.setCurrentSelectedDayBackgroundColor(currentSelectedDayBackgroundColor)
        invalidate()
    }

    fun setCurrentDayBackgroundColor(currentDayBackgroundColor: Int) {
        compactCalendarController.setCurrentDayBackgroundColor(currentDayBackgroundColor)
        invalidate()
    }

    val heightPerDay: Int
        get() = compactCalendarController.heightPerDay

    fun setListener(listener: CompactCalendarViewListener?) {
        compactCalendarController.setListener(listener)
    }

    val firstDayOfCurrentMonth: Date
        get() = compactCalendarController.firstDayOfCurrentMonth

    fun shouldDrawIndicatorsBelowSelectedDays(shouldDrawIndicatorsBelowSelectedDays: Boolean) {
        compactCalendarController.shouldDrawIndicatorsBelowSelectedDays(shouldDrawIndicatorsBelowSelectedDays)
    }

    fun setCurrentDate(dateTimeMonth: Date) {
        compactCalendarController.setCurrentDate(dateTimeMonth)
        invalidate()
    }

    val weekNumberForCurrentMonth: Int
        get() = compactCalendarController.weekNumberForCurrentMonth

    fun setShouldDrawDaysHeader(shouldDrawDaysHeader: Boolean) {
        compactCalendarController.setShouldDrawDaysHeader(shouldDrawDaysHeader)
    }

    fun setCurrentSelectedDayTextColor(currentSelectedDayTextColor: Int) {
        compactCalendarController.setCurrentSelectedDayTextColor(currentSelectedDayTextColor)
    }

    fun setCurrentDayTextColor(currentDayTextColor: Int) {
        compactCalendarController.setCurrentDayTextColor(currentDayTextColor)
    }
    /**
     * Adds an event to be drawn as an indicator in the calendar.
     * If adding multiple events see [.addEvents]} method.
     * @param event to be added to the calendar
     * @param shouldInvalidate true if the view should invalidate
     *
     * see [.addEvent] when adding single events to control if calendar should redraw
     * or [.addEvents]  when adding multiple events
     * @param event
     */
    @JvmOverloads
    fun addEvent(event: Event, shouldInvalidate: Boolean = true) {
        compactCalendarController.addEvent(event)
        if (shouldInvalidate) {
            invalidate()
        }
    }

    /**
     * Adds multiple events to the calendar and invalidates the view once all events are added.
     */
    fun addEvents(events: List<Event>) {
        compactCalendarController.addEvents(events)
        invalidate()
    }

    /**
     * Fetches the events for the date passed in
     * @param date
     * @return
     */
    fun getEvents(date: Date): List<Event?>? {
        return compactCalendarController.getCalendarEventsFor(date.time)
    }

    /**
     * Fetches the events for the epochMillis passed in
     * @param epochMillis
     * @return
     */
    fun getEvents(epochMillis: Long): List<Event?>? {
        return compactCalendarController.getCalendarEventsFor(epochMillis)
    }

    /**
     * Fetches the events for the month of the epochMillis passed in and returns a sorted list of events
     * @param epochMillis
     * @return
     */
    fun getEventsForMonth(epochMillis: Long): List<Event?>? {
        return compactCalendarController.getCalendarEventsForMonth(epochMillis)
    }

    /**
     * Fetches the events for the month of the date passed in and returns a sorted list of events
     * @param date
     * @return
     */
    fun getEventsForMonth(date: Date): List<Event> {
        return compactCalendarController.getCalendarEventsForMonth(date.time)
    }

    /**
     * Remove the event associated with the Date passed in
     * @param date
     */
    fun removeEvents(date: Date) {
        compactCalendarController.removeEventsFor(date.time)
    }

    fun removeEvents(epochMillis: Long) {
        compactCalendarController.removeEventsFor(epochMillis)
    }
    /**
     * Removes an event from the calendar.
     * If removing multiple events see [.removeEvents]
     *
     * @param event event to remove from the calendar
     * @param shouldInvalidate true if the view should invalidate
     *
     * see [.removeEvent] when removing single events to control if calendar should redraw
     * or [.removeEvents] (java.util.List)}  when removing multiple events
     * @param event
     */
    @JvmOverloads
    fun removeEvent(event: Event, shouldInvalidate: Boolean = true) {
        compactCalendarController.removeEvent(event)
        if (shouldInvalidate) {
            invalidate()
        }
    }

    /**
     * Removes multiple events from the calendar and invalidates the view once all events are added.
     */
    fun removeEvents(events: List<Event>) {
        compactCalendarController.removeEvents(events)
        invalidate()
    }

    /**
     * Clears all Events from the calendar.
     */
    fun removeAllEvents() {
        compactCalendarController.removeAllEvents()
        invalidate()
    }

    fun setIsRtl(isRtl: Boolean) {
        compactCalendarController.setIsRtl(isRtl)
    }

    fun shouldSelectFirstDayOfMonthOnScroll(shouldSelectFirstDayOfMonthOnScroll: Boolean) {
        compactCalendarController.setShouldSelectFirstDayOfMonthOnScroll(shouldSelectFirstDayOfMonthOnScroll)
    }

    fun setCurrentSelectedDayIndicatorStyle(currentSelectedDayIndicatorStyle: Int) {
        compactCalendarController.setCurrentSelectedDayIndicatorStyle(currentSelectedDayIndicatorStyle)
        invalidate()
    }

    fun setCurrentDayIndicatorStyle(currentDayIndicatorStyle: Int) {
        compactCalendarController.setCurrentDayIndicatorStyle(currentDayIndicatorStyle)
        invalidate()
    }

    fun setEventIndicatorStyle(eventIndicatorStyle: Int) {
        compactCalendarController.setEventIndicatorStyle(eventIndicatorStyle)
        invalidate()
    }

    private fun checkTargetHeight() {
        check(compactCalendarController.targetHeight > 0) { "Target height must be set in xml properties in order to expand/collapse CompactCalendar." }
    }

    fun displayOtherMonthDays(displayOtherMonthDays: Boolean) {
        compactCalendarController.setDisplayOtherMonthDays(displayOtherMonthDays)
        invalidate()
    }

    fun setTargetHeight(targetHeight: Int) {
        compactCalendarController.targetHeight = targetHeight
        checkTargetHeight()
    }

    fun showCalendar() {
        checkTargetHeight()
        animationHandler.openCalendar()
    }

    fun hideCalendar() {
        checkTargetHeight()
        animationHandler.closeCalendar()
    }

    fun showCalendarWithAnimation() {
        checkTargetHeight()
        animationHandler.openCalendarWithAnimation()
    }

    fun hideCalendarWithAnimation() {
        checkTargetHeight()
        animationHandler.closeCalendarWithAnimation()
    }

    /**
     * Moves the calendar to the right. This will show the next month when [.setIsRtl]
     * is set to false. If in rtl mode, it will show the previous month.
     */
    fun scrollRight() {
        compactCalendarController.scrollRight()
        invalidate()
    }

    /**
     * Moves the calendar to the left. This will show the previous month when [.setIsRtl]
     * is set to false. If in rtl mode, it will show the next month.
     */
    fun scrollLeft() {
        compactCalendarController.scrollLeft()
        invalidate()
    }

    val isAnimating: Boolean
        get() = animationHandler.isAnimating

    override fun onMeasure(parentWidth: Int, parentHeight: Int) {
        super.onMeasure(parentWidth, parentHeight)
        val width = MeasureSpec.getSize(parentWidth)
        val height = MeasureSpec.getSize(parentHeight)
        if (width > 0 && height > 0) {
            compactCalendarController.onMeasure(width, height, paddingRight, paddingLeft)
        }
        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        compactCalendarController.onDraw(canvas)
    }

    override fun computeScroll() {
        super.computeScroll()
        if (compactCalendarController.computeScroll()) {
            invalidate()
        }
    }

    fun shouldScrollMonth(enableHorizontalScroll: Boolean) {
        horizontalScrollEnabled = enableHorizontalScroll
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (horizontalScrollEnabled) {
            compactCalendarController.onTouch(event)
            invalidate()
        }
        // on touch action finished (CANCEL or UP), we re-allow the parent container to intercept touch events (scroll inside ViewPager + RecyclerView issue #82)
        if ((event.action == MotionEvent.ACTION_CANCEL || event.action == MotionEvent.ACTION_UP) && horizontalScrollEnabled) {
            parent.requestDisallowInterceptTouchEvent(false)
        }
        // always allow gestureDetector to detect onSingleTap and scroll events
        return gestureDetector.onTouchEvent(event)
    }

    override fun canScrollHorizontally(direction: Int): Boolean {
        return if (this.visibility == GONE) {
            false
        } else horizontalScrollEnabled
        // Prevents ViewPager from scrolling horizontally by announcing that (issue #82)
    }

    companion object {
        const val FILL_LARGE_INDICATOR = 1
        const val NO_FILL_LARGE_INDICATOR = 2
        const val SMALL_INDICATOR = 3
    }

    init {
        gestureDetector = GestureDetectorCompat(getContext(), gestureListener)
        animationHandler = AnimationHandler(compactCalendarController, this)
    }
}
